blog/2022-06-21-using-temporal-as-a-node-task-queue.md

```diff
@@ -0,0 +1,372 @@
+
+```ts
+import axios from 'axios';
+import { SendEmailParams } from './interfaces';
+
+interface MailgunSettings {
+  apiKey: string;
+  domain: string;
+}
+
+const mailgunAPI = 'https://api.mailgun.net/v3';
+
+export const createActivities = ({ apiKey, domain }: MailgunSettings) => ({
+  async sendEmail(params: SendEmailParams): Promise<void> {
+    const { to, from, subject, html } = params;
+    await axios({
+      url: `${mailgunAPI}/${domain}/messages`,
+      method: 'post',
+      params: { to, from, subject, html },
+      auth: {
+        username: 'api',
+        password: apiKey,
+      },
+    });
+  },
+});
+```
+
+Next, you need to define a Workflow that calls this activity.
+
+`src/workflows.ts`
+
+```ts
+import { proxyActivities } from '@temporalio/workflow';
+import type { createActivities } from './activities';
+import { SendEmailParams } from './interfaces';
+
+const { sendEmail } = proxyActivities<ReturnType<typeof createActivities>>({
+  startToCloseTimeout: '5m',
+});
+
+export async function sendEmailWorkflow(params: SendEmailParams): Promise<void> {
+  await sendEmail(params);
+}
+```
+
+Finally, you need to create [Worker](https://docs.temporal.io/concepts/what-is-a-worker-entity) to process tasks as follows.
+
+`src/worker.ts`
+
+```ts
+import { Worker } from '@temporalio/worker';
+import dotenv from 'dotenv';
+import { createActivities } from './activities';
+
+dotenv.config();
+
+const {
+  MAILGUN_API: apiKey,
+  MAILGUN_DOMAIN: domain
+} = process.env;
+
+async function run(): Promise<void> {
+  if (apiKey === undefined) throw new Error('No API key');
+  if (domain === undefined) throw new Error('No domain');
+
+  const activities = createActivities({
+    apiKey,
+    domain,
+  });
+
+  const worker = await Worker.create({
+    taskQueue: 'email',
+    activities,
+    workflowsPath: require.resolve('./workflows'),
+  });
+  await worker.run();
+}
+
+run().catch((err) => {
+  console.error(err);
+  process.exit(1);
+});
+```
+
+Finally, to start a Workflow, you need to create a `WorkflowClient` and `start()` the Workflow as follows.
+
+`src/client.ts`
+
+```ts
+import { Connection, WorkflowClient } from '@temporalio/client';
+import { sendEmailWorkflow } from './workflows';
+import { nanoid } from 'nanoid';
+
+const connection = new Connection({});
+const client = new WorkflowClient(connection.service);
+
+const handle = await client.start(sendEmailWorkflow, {
+  args: [{
+    to: 'john.smith@gmail.com',
+    from: 'test@temporal.io',
+    subject: 'Welcome to Temporal!',
+    html: 'This is a test email'
+  }], // type inference works! args: [name: string]
+  taskQueue: 'email',
+  // in practice, use a meaningful business id, eg customerId or transactionId
+  workflowId: 'workflow-' + nanoid(),
+});
+```
+
+The Temporal code is slight more verbose, because you need a separate Activity function.
+However, Temporal has several features that make it an excellent option for the task queue use case.
```

Comment:
I'd also note that in contrast with the queue example, activity and workflow invocations are type checked.